src/pkg/artifact/manager/dao/model.go

```diff
@@ -0,0 +1,173 @@
+// Copyright Project Harbor Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dao
+
+import (
+	"encoding/json"
+	"time"
+
+	"github.com/astaxie/beego/orm"
+	"github.com/goharbor/harbor/src/common/models"
+	"github.com/goharbor/harbor/src/common/utils/log"
+	"github.com/goharbor/harbor/src/pkg/artifact/model"
+)
+
+func init() {
+	orm.RegisterModel(&Artifact{})
+	orm.RegisterModel(&Tag{})
+	orm.RegisterModel(&ArtifactReference{})
+}
+
+// Artifact model in database
+type Artifact struct {
+	ID                int64     `orm:"pk;auto;column(id)"`
+	Type              string    `orm:"column(type)"`                // image or chart
+	MediaType         string    `orm:"column(media_type)"`          // the media type of artifact
+	ManifestMediaType string    `orm:"column(manifest_media_type)"` // the media type of manifest/index
+	ProjectID         int64     `orm:"column(project_id)"`          // needed for quota
+	RepositoryID      int64     `orm:"column(repository_id)"`
+	Digest            string    `orm:"column(digest)"`
+	Size              int64     `orm:"column(size)"`
+	PushTime          time.Time `orm:"column(push_time)"`
+	Platform          string    `orm:"column(platform)"`    // json string
+	ExtraAttrs        string    `orm:"column(extra_attrs)"` // json string
+	Annotations       string    `orm:"column(annotations)"` // json string
+	Revision          int64     `orm:"column(revision)"`    // record data revision
+}
+
+// TableName for artifact
+func (a *Artifact) TableName() string {
+	// TODO use "artifact" after finishing the upgrade/migration work
+	return "artifact_2"
+}
+
+// To converts the artifact to the business level object
+func (a *Artifact) To() *model.Artifact {
```

Comment:
One more comment here: with this `To` function, lower package will refer to the top-level package? Maybe the from convert function can be defined in the top-level package?

Reply:
Updated